### DISCUSSION TITLE
DuckDB exhibits inconsistent behavior when reading from S3

### DISCUSSION THREAD

**Discussion by sriramr98** [duckdb/duckdb#15480] [Q&A]
I have a FastAPI service which reads some data from parquet files in S3 before processing that information. When the service is deployed, the code works exactly as expected for 6-8 hours and then starts throwing a 400 error on all reads. As soon as the service is restarted, the code works again and DuckDB is able to read from S3 for the next 6-8 hours before exhibiting the same behavior of throwing 400 errors on reads

Here's a minimally reproducible code

`app.py`
```python
from fastapi import FastAPI, Response, status
from dotenv import load_dotenv
from pydantic import BaseModel

from s3 import fetchDataFromS3, init_duckdb

load_dotenv()

init_duckdb()

app = FastAPI()

class DataPayload(BaseModel):
    path: str

@app.post("/data")
def getData(data: DataPayload, response: Response):
    try:
        resp = fetchDataFromS3(data.path)

        if resp["success"]:
            return {"success": True, "body": resp["data"]}
        else:
            response.status_code = status.HTTP_400_BAD_REQUEST
            return {"success": False, "error": resp["error"]} 
    except Exception as e:
        print(e)
        response.status_code = status.HTTP_500_INTERNAL_SERVER_ERROR
        return {"success": False, "error": {"errCode": "E106", "message": str(e)}}
```

`s3.py`
```python
from botocore.config import Config
import boto3

import duckdb
import os

s3_config = Config(read_timeout=30)
s3 = boto3.client('s3', config=s3_config)

db = duckdb.connect(database=':memory:')

def fetchDataFromS3(filePath: str):
    try:
        BUCKET_NAME = os.getenv('S3_READ_BUCKET') or ''

        query = f"SELECT * FROM 's3://{BUCKET_NAME}/{filePath}';";
        resp = db.sql(query)

        data = resp.pl().rows(named=True)
        return {
            'success': True,
            'data': data
        }
    except:
        return {
            'success': False,
            'error': 'Unable to read data from S3'
        }



def init_duckdb():
    db.execute('pragma autoload_known_extensions=true')
    db.execute("SET extension_directory = '/tmp/duckdb';")
    db.execute(f"CREATE SECRET IF NOT EXISTS secret2 (TYPE S3,PROVIDER CREDENTIAL_CHAIN, REGION '{os.getenv('REGION')}');")
```

`.env`
```
S3_READ_BUCKET=<your_bucket_name>
REGION=<your_bucket_region>
```

`requirements.txt` ( this is a fresh env created, so most of these are transitive dependencies )
```
annotated-types==0.7.0
anyio==4.7.0
boto3==1.35.90
botocore==1.35.90
certifi==2024.12.14
click==8.1.8
dnspython==2.7.0
duckdb==1.1.3
email_validator==2.2.0
fastapi==0.115.6
fastapi-cli==0.0.7
h11==0.14.0
httpcore==1.0.7
httptools==0.6.4
httpx==0.28.1
idna==3.10
Jinja2==3.1.5
jmespath==1.0.1
markdown-it-py==3.0.0
MarkupSafe==3.0.2
mdurl==0.1.2
polars==1.18.0
pyarrow==18.1.0
pydantic==2.10.4
pydantic_core==2.27.2
Pygments==2.18.0
python-dateutil==2.9.0.post0
python-dotenv==1.0.1
python-multipart==0.0.20
PyYAML==6.0.2
rich==13.9.4
rich-toolkit==0.12.0
s3transfer==0.10.4
shellingham==1.5.4
six==1.17.0
sniffio==1.3.1
starlette==0.41.3
typer==0.15.1
typing_extensions==4.12.2
urllib3==2.3.0
uvicorn==0.34.0
uvloop==0.21.0
watchfiles==1.0.3
websockets==14.1
```

If you run this service and simulate requests for a period of time, after 6-8 hours, the S3 reads start throwing a 400 error ( screenshot below )

<img width="870" alt="Screenshot 2024-12-27 at 3 30 20 PM" src="https://github.com/user-attachments/assets/2e112046-c1c8-490b-9eb3-b7cc1fd11125" />

However, as soon as I restart the service inside the VM, the reads start working again without any changes at all.

---

## FIXES TRIED

With a small modification to `init_duckdb()` function and `fetchDataFromS3` function, I was able to make it work without any issues, but the reason why something like this should work boggles me. 

### MODIFIED CODE

`s3.py`

```python
s3_config = Config(read_timeout=30)
s3 = boto3.client('s3', config=s3_config)

db = duckdb.connect(database=':memory:')

def init_duckdb():
    global db, s3

    # Close existing connection if any
    try:
        db.close()
    except:
        pass

    # Create fresh boto3 client to refresh credentials
    s3_config = Config(read_timeout=30)
    s3 = boto3.client('s3', config=s3_config)

    # Create fresh DuckDB connection
    db = duckdb.connect(database=':memory:')

    logger.warning("Starting DuckDB initialization")
    db.execute('pragma autoload_known_extensions=true')
    db.execute("SET extension_directory = '/tmp/duckdb';")
    db.execute("DROP SECRET IF EXISTS secret2")  # Force drop existing secret
    db.execute(f"CREATE SECRET IF NOT EXISTS secret2 (TYPE S3,PROVIDER CREDENTIAL_CHAIN, REGION '{os.getenv('REGION')}');")  # Force create new secret
    logger.warning("DuckDB initialization completed successfully")


def fetchDataFromS3(filePath: str):
    init_duckdb()
    # Same code as before, ignored for brevity
```

Essentially, I'm destroying the secret and re-creating it for every READ request sent to S3. After making this change, I've not seen the issue pop up for days whereas earlier, we see it in every read until service restarts ( manual intervention ) and then issue starts again in some hours.

Why does  dropping and re-creating the secret fix this? Is there a time limit or expiry to the secrets that are created?

I haven't raised a bug for this because as of now, I do not know if this is a bug or if there's something I'm missing.

> NOTE: If you're trying to reproduce the issue, you will have to run the service somewhere for a couple of hours and keep trying it at intervals. It's very inconsistent on when it pops up, but it usually shows up after about 6-8 hours of running the service. As soon as the service is restarted, the issue disappears for a couple hours before showing up again.

**Comment by soerenwolfers** [comment]
What's the point of using `boto` here? DuckDB has the credentials fetching functionality built-in. 

Also, could you show the actual error duckdb is throwing? Screenshots of your *own* error wrappers aren't very helpful.

Finally, and I'm far from an expert on those but I think it's plausible that credentials expire and DuckDB should be able to auto-refresh them but isn't. 

If you want people to look at this, you should make a proper issue out of this and include an actual *Minimal* Working Example. From the looks of it, it should be possible to make this MWE pure-SQL, i.e. drop all Python and provide something that could be pasted into a DuckDB CLI instance.

In short, your issue should look something like:

> **What happens?**
> If I run an S3 query more than 8 hours after secret creation, I get the error 
> ```
><Error>
>```
>
> **To reproduce**: 
> Run the following in a CLI instance, including the waiting period indicated in the comment. 
>```sql
>CREATE SECRET secret2 (TYPE S3,PROVIDER CREDENTIAL_CHAIN, REGION '<region>');
> --Wait 8 Hours
>SELECT * FROM 's3://<path>'
>```
> Relevant information about my S3 setup: <e.g., what's in your ~/.aws, relevant environment variables, etc.>
